design/design.md

```diff
@@ -0,0 +1,380 @@
+order to post more coroutines corresponding to additional requests.
+`CompletionQueue` is thread safe.
+
+The `ResourceManager` of rippled will be used to track resource consumption per
+endpoint. If the `ResourceManager` indicates that we should "disconnect" from
+a specific endpoint, a coroutine will not be posted to the `JobQueue`, the
+handler will not be called and an error will be returned to the client.
+
+## Classes
+The server code is adapted from the example found
+[here](https://github.com/grpc/grpc/blob/v1.24.0/examples/cpp/helloworld/greeter_async_server.cc)
+
+`CompletionQueue` is a class of the GRPC library. `CompletionQueue` is used to
+queue events, and provides a method, `CompletionQueue::Next()` to query for any
+events that have occurred. Events are returned by `CompletionQueue` in the form
+of a `void **` pointer, known as a tag. In our application, `CompletionQueue` returns pointers to
+`CallData` objects, which are used to service requests.
+
+Each request is represented by a `CallData` object. `CallData` is an abstract
+class that implements a method `Proceed()` and can be in one of three states:
+LISTEN, PROCESS or FINISH. These states correspond to different states in the
+lifecycle of an rpc call, and `Proceed()` transitions the object from one state
+to the next. `CallData` has two pure virtual methods: `makeListener()` and
+`process()`, which are implemented by derived classes. For
+each rpc method, there is a class that derives from `CallData`, and implements
+the pure virtual methods. `finish()` can be overridden if necessary. The
+destructor is virtual so that `delete this` actually deletes the entire object,
+including the derived class.
+
+**In the below example, some code is ommitted for the sake of clarity and
+brevity, including arguments to functions and member variables**
+
+```
+class CallData {
+
+    CallData(CompletionQueue* cq) :cq_(cq), _status(LISTEN)
+    {
+    }
+
+    virtual ~CallData() {}
+    void Proceed() {
+      if (status_ == LISTEN) {
+          status_ = PROCESS;
+          process();
+      } else {
+          _status = FINISH;
+          finish();
+          delete this;
+      }
+    }
+
+    virtual void makeListener() = 0;
+    virtual void process() = 0;
+    virtual void finish() {};
+
+    CompletionQueue* cq_;
+    enum CallStatus { LISTEN, PROCESS, FINISH};
+    CallStatus status_;
+};
+
+class GetAccountInfoCallData : public CallData {
+
+    AccountInfoCallData(CompletionQueue* cq) : CallData(cq)
+    {
+        makeListener();
+    }
+
+    void makeListener()
+    {
+        // We *request* that the system start listening for GetAccountInfo
+        // requests. "this" acts as the tag uniquely identifying the request 
+        // (so that different CallData instances can serve different requests
+        // concurrently), in this case the memory address of this CallData instance.
+        service_->RequestGetAccountInfo(&ctx_, &request_, &responder_, cq_, cq_, this);
```

Comment:
I think I missed what `service_` is. Where did this come from?

Reply:
I left the declaration of that member out of the doc for brevity, but its a class that is codegen'd by GRPC, and is a way to interface with the service. This call here tells the service to listen for the next GetAccountInfo request.